File: Examples/Diagram/WebForm/VB.NET/Examples/WebControl/Layouts/NOrthogonalGraphLayoutUC.ascx.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Windows.Forms

Imports Nevron.Dom
Imports Nevron.Diagram
Imports Nevron.Diagram.Filters
Imports Nevron.Diagram.Batches
Imports Nevron.Diagram.DataStructures
Imports Nevron.Diagram.Layout
Imports Nevron.Diagram.Templates
Imports Nevron.Diagram.Shapes
Imports Nevron.Diagram.WebForm
Imports Nevron.UI.WebForm.Controls
Imports Nevron.GraphicsCore

Namespace Nevron.Examples.Diagram.Webform
	''' <summary>
	'''	Summary description for NOrthogonalGraphLayoutUC.
	''' </summary>
	Public Partial Class NOrthogonalGraphLayoutUC
		Inherits NDiagramExampleUC
		Protected Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs)
			If NDrawingView1.RequiresInitialization Then
				' begin view init
				NDrawingView1.ViewLayout = CanvasLayout.Fit
				NDrawingView1.GlobalVisibility.ShowPorts = False
				NDrawingView1.GlobalVisibility.ShowArrowheads = False

				' init document
				NDrawingView1.Document.HistoryService.Stop()
				NDrawingView1.Document.BeginInit()
				InitDocument()
				NDrawingView1.Document.EndInit()
			End If
		End Sub

		#Region "Implementation"

		Private Sub InitDocument()
			Dim document As NDrawingDocument = NDrawingView1.Document

			' remove the standard frame
			document.BackgroundStyle.FrameStyle.Visible = False

			' adjust the graphics quality
			document.GraphicsSettings.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighSpeed

			' set up visual formatting
			document.Style.FillStyle = New NGradientFillStyle(GradientStyle.Horizontal, GradientVariant.Variant3, Color.FromArgb(192, 194, 194), Color.FromArgb(129, 133, 133))
			document.Style.StrokeStyle = New NStrokeStyle(1, Color.FromArgb(68, 90, 108))

			Dim sheet As NStyleSheet = New NStyleSheet("edges")
			sheet.Style.StartArrowheadStyle = New NArrowheadStyle(ArrowheadShape.None, "", NSizeL.Empty, document.Style.FillStyle, document.Style.StrokeStyle)
			sheet.Style.EndArrowheadStyle = New NArrowheadStyle(ArrowheadShape.None, "", NSizeL.Empty, document.Style.FillStyle, document.Style.StrokeStyle)
			document.StyleSheets.AddChild(sheet)

			' create a tree
			CreateDiagram()

			' perform the layout
			PerformLayout(Nothing)
		End Sub
		Private Sub CreateDiagram()
			Const width As Integer = 100, height As Integer = 100, distance As Integer = 200
			Dim f As NBasicShapesFactory = New NBasicShapesFactory(NDrawingView1.Document)
			Dim edge As NRoutableConnector
			Dim from As Integer() = New Integer() { 1, 1, 1, 2, 2, 2, 3, 3, 4, 4, 4, 5, 5, 6 }
			Dim [to] As Integer() = New Integer() { 2, 3, 4, 4, 5, 8, 6, 7, 5, 8, 10, 8, 9, 10 }
			Dim shapes As NShape() = New NShape(9){}
			Dim vertexCount As Integer = shapes.Length
			Dim edgeCount As Integer = from.Length
			Dim count As Integer = vertexCount + edgeCount

			Dim i As Integer = 0
			Do While i < count
				If i < vertexCount Then
					Dim j As Integer
					If vertexCount Mod 2 = 0 Then
						j = i
					Else
						j = i + 1
					End If
					shapes(i) = f.CreateShape(CInt(Fix(BasicShapes.Rectangle)))

					If vertexCount Mod 2 <> 0 AndAlso i = 0 Then
						shapes(i).Bounds = New NRectangleF(CInt(Fix(width + (distance * 1.5))) / 2, distance + (j / 2) * CInt(Fix(distance * 1.5)), width, height)
					Else
						shapes(i).Bounds = New NRectangleF(width / 2 + (j Mod 2) * CInt(Fix(distance * 1.5)), height + (j / 2) * CInt(Fix(distance * 1.5)), width, height)
					End If

					NDrawingView1.Document.ActiveLayer.AddChild(shapes(i))
				Else
					edge = New NRoutableConnector()
					edge.ConnectorType = RoutableConnectorType.DynamicPolyline
					edge.StyleSheetName = "edges"
					NDrawingView1.Document.ActiveLayer.AddChild(edge)
					edge.FromShape = shapes(from(i - vertexCount) - 1)
					edge.ToShape = shapes([to](i - vertexCount) - 1)
				End If
				i += 1
			Loop
		End Sub
		Private Sub CreateGraph(ByVal vertexCount As Integer, ByVal edgeCount As Integer)
			NDrawingView1.Document.ActiveLayer.RemoveAllChildren()

			Dim randomGraph As NRandomGraphTemplate = New NRandomGraphTemplate()
			randomGraph.EdgesStyleSheetName = "edges"
			randomGraph.VertexCount = vertexCount
			randomGraph.EdgeCount = edgeCount
			randomGraph.Create(NDrawingView1.Document)
		End Sub
		Private Sub PerformLayout(ByVal args As Hashtable)
			' Create the layout
			Dim layout As NOrthogonalGraphLayout = New NOrthogonalGraphLayout()

			' Configure the layout
			NLayoutsHelper.ConfigureLayout(layout, args)

			' Get the shapes to layout
			Dim shapes As NNodeList = NDrawingView1.Document.ActiveLayer.Children(NFilters.Shape2D)

			' Layout the shapes
			layout.Layout(shapes, New NDrawingLayoutContext(NDrawingView1.Document))

			' Resize document to fit all shapes
			NDrawingView1.Document.SizeToContent()
		End Sub

		#End Region 

		#Region "Event Handlers"

		Protected Sub NDrawingView1_AsyncCustomCommand(ByVal sender As Object, ByVal e As EventArgs)
			Dim args As NCallbackCustomCommandArgs = TryCast(e, NCallbackCustomCommandArgs)
			Select Case args.Command.Name
				Case "btn1"
					CreateGraph(10, 15)

				Case "btn2"
					CreateGraph(20, 25)
			End Select

			PerformLayout(args.Command.Arguments)
			m_bClientSideRedrawRequired = True
		End Sub
		Protected Sub NDrawingView1_AsyncQueryCommandResult(ByVal sender As Object, ByVal e As EventArgs)
			Dim args As NCallbackQueryCommandResultArgs = TryCast(e, NCallbackQueryCommandResultArgs)
			Dim resultBuilder As NAjaxXmlTransportBuilder = args.ResultBuilder
			If m_bClientSideRedrawRequired AndAlso (Not resultBuilder.ContainsRedrawDataSection()) Then
				resultBuilder.AddRedrawDataSection(NDrawingView1)
			End If
		End Sub

		#End Region

		#Region "Fields"

		Private m_bClientSideRedrawRequired As Boolean = False

		#End Region
	End Class
End Namespace

File: Examples/Diagram/WebForm/CSharp/Examples/WebControl/Layouts/NOrthogonalGraphLayoutUC.ascx.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Windows.Forms;

using Nevron.Dom;
using Nevron.Diagram;
using Nevron.Diagram.Filters;
using Nevron.Diagram.Batches;
using Nevron.Diagram.DataStructures;
using Nevron.Diagram.Layout;
using Nevron.Diagram.Templates;
using Nevron.Diagram.Shapes;
using Nevron.Diagram.WebForm;
using Nevron.UI.WebForm.Controls;
using Nevron.GraphicsCore;

namespace Nevron.Examples.Diagram.Webform
{
	/// <summary>
	///	Summary description for NOrthogonalGraphLayoutUC.
	/// </summary>
	public partial class NOrthogonalGraphLayoutUC : NDiagramExampleUC
	{
		protected void Page_Load(object sender, System.EventArgs e)
		{
            if (NDrawingView1.RequiresInitialization)
            {
                // begin view init
                NDrawingView1.ViewLayout = CanvasLayout.Fit;
                NDrawingView1.GlobalVisibility.ShowPorts = false;
                NDrawingView1.GlobalVisibility.ShowArrowheads = false;

                // init document
                NDrawingView1.Document.HistoryService.Stop();
                NDrawingView1.Document.BeginInit();
                InitDocument();
                NDrawingView1.Document.EndInit();
            }
        }

        #region Implementation

        private void InitDocument()
		{
            NDrawingDocument document = NDrawingView1.Document;

            // remove the standard frame
            document.BackgroundStyle.FrameStyle.Visible = false;

            // adjust the graphics quality
            document.GraphicsSettings.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighSpeed;

            // set up visual formatting
            document.Style.FillStyle = new NGradientFillStyle(GradientStyle.Horizontal, GradientVariant.Variant3, Color.FromArgb(192, 194, 194), Color.FromArgb(129, 133, 133));
            document.Style.StrokeStyle = new NStrokeStyle(1, Color.FromArgb(68, 90, 108));

            NStyleSheet sheet = new NStyleSheet("edges");
            sheet.Style.StartArrowheadStyle = new NArrowheadStyle(ArrowheadShape.None, "", NSizeL.Empty,
                document.Style.FillStyle, document.Style.StrokeStyle);
            sheet.Style.EndArrowheadStyle = new NArrowheadStyle(ArrowheadShape.None, "", NSizeL.Empty,
                document.Style.FillStyle, document.Style.StrokeStyle);
            document.StyleSheets.AddChild(sheet);

            // create a tree
            CreateDiagram();

            // perform the layout
            PerformLayout(null);
        }
		private void CreateDiagram()
		{
			const int width = 100, height = 100, distance = 200;
			NBasicShapesFactory f = new NBasicShapesFactory(NDrawingView1.Document);
			NRoutableConnector edge;
			int[] from = new int[] { 1, 1, 1, 2, 2, 2, 3, 3, 4, 4, 4, 5, 5, 6 };
			int[] to = new int[] { 2, 3, 4, 4, 5, 8, 6, 7, 5, 8, 10, 8, 9, 10 };
			NShape[] shapes = new NShape[10];
			int vertexCount = shapes.Length;
			int edgeCount = from.Length;
			int count = vertexCount + edgeCount;

			for (int i = 0; i < count; i++)
			{
				if (i < vertexCount)
				{
					int j = vertexCount % 2 == 0 ? i : i + 1;
					shapes[i] = f.CreateShape((int)BasicShapes.Rectangle);

					if (vertexCount % 2 != 0 && i == 0)
					{
						shapes[i].Bounds = new NRectangleF((int)(width + (distance * 1.5)) / 2,
							distance + (j / 2) * (int)(distance * 1.5), width, height);
					}
					else
					{
						shapes[i].Bounds = new NRectangleF(width / 2 + (j % 2) * (int)(distance * 1.5),
							height + (j / 2) * (int)(distance * 1.5), width, height);
					}

					NDrawingView1.Document.ActiveLayer.AddChild(shapes[i]);
				}
				else
				{
					edge = new NRoutableConnector();
					edge.ConnectorType = RoutableConnectorType.DynamicPolyline;
					edge.StyleSheetName = "edges";
					NDrawingView1.Document.ActiveLayer.AddChild(edge);
					edge.FromShape = shapes[from[i - vertexCount] - 1];
					edge.ToShape = shapes[to[i - vertexCount] - 1];
				}
			}
		}
        private void CreateGraph(int vertexCount, int edgeCount)
        {
            NDrawingView1.Document.ActiveLayer.RemoveAllChildren();

            NRandomGraphTemplate randomGraph = new NRandomGraphTemplate();
            randomGraph.EdgesStyleSheetName = "edges";
            randomGraph.VertexCount = vertexCount;
            randomGraph.EdgeCount = edgeCount;
            randomGraph.Create(NDrawingView1.Document);
        }
		private void PerformLayout(Hashtable args)
		{
			// Create the layout
			NOrthogonalGraphLayout layout = new NOrthogonalGraphLayout();

            // Configure the layout
            NLayoutsHelper.ConfigureLayout(layout, args);

            // Get the shapes to layout
            NNodeList shapes = NDrawingView1.Document.ActiveLayer.Children(NFilters.Shape2D);

            // Layout the shapes
			layout.Layout(shapes, new NDrawingLayoutContext(NDrawingView1.Document));

            // Resize document to fit all shapes
            NDrawingView1.Document.SizeToContent();
        }

        #endregion 

        #region Event Handlers

        protected void NDrawingView1_AsyncCustomCommand(object sender, EventArgs e)
        {
            NCallbackCustomCommandArgs args = e as NCallbackCustomCommandArgs;
            switch (args.Command.Name)
            {
                case "btn1":
                    CreateGraph(10, 15);
                    break;

                case "btn2":
                    CreateGraph(20, 25);
                    break;
            }

            PerformLayout(args.Command.Arguments);
            m_bClientSideRedrawRequired = true;
        }
        protected void NDrawingView1_AsyncQueryCommandResult(object sender, EventArgs e)
        {
            NCallbackQueryCommandResultArgs args = e as NCallbackQueryCommandResultArgs;
            NAjaxXmlTransportBuilder resultBuilder = args.ResultBuilder;
            if (m_bClientSideRedrawRequired && !resultBuilder.ContainsRedrawDataSection())
            {
                resultBuilder.AddRedrawDataSection(NDrawingView1);
            }
        }

        #endregion

        #region Fields

        private bool m_bClientSideRedrawRequired = false;

        #endregion
    }
}